### PR TITLE
Add Vectorized<c10::BFloat16> specialization for ARM

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139220
* #139208
* #139081
* #139558
* __->__ #139090
* #139084

When we have hardware support, we can use it. When we don't have hardware support, we can still do better than vec_base.h. I'm not sure to what extent we're set up to properly test both `defined(__ARM_FEATURE_BF16)` and `!defined(__ARM_FEATURE_BF16)` builds, feedback especially welcome there.

Testing: vec_test_all_types should cover correctness. For perf, seems clear that using vectorized intrinsics should be better than vec_base?

Differential Revision: [D64997747](https://our.internmc.facebook.com/intern/diff/D64997747/)

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by swolchok** [comment]
the inductor failures look like this just needs a re-export, but there is a legit-looking mac failure as wel

File: aten/src/ATen/cpu/vec/vec128/vec128_bfloat16_neon.h [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+
+// Unlike the float16_t family of types, bfloat16_t is not available
+// when we're not targeting bfloat16 hardware support. (See
+// https://godbolt.org/z/orv6e94n4 ) So, we need to handle it at
+// uint16_t in that case.
+#ifndef __ARM_FEATURE_BF16
+using bfloat16x8_t = uint16x8_t;
+using bfloat16x4_t = uint16x4_t;
+using bfloat16_t = uint16_t;
+#define vsetq_lane_bf16 vsetq_lane_u16
```
Comment (swolchok) [inline aten/src/ATen/cpu/vec/vec128/vec128_bfloat16_neon.h:31]: this macro-based approach seems problematic; the Mac SDK defines these macros even when not compiling with __ARM_FEATURE_BF16. I'll have to figure out something else.

Comment (swolchok) [reply aten/src/ATen/cpu/vec/vec128/vec128_bfloat16_neon.h:31]: specifically the problem is leaking the macros into other translation units.

File: aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h [hunk 2]
```diff
@@ -105,6 +105,25 @@ class Vectorized<c10::Half> : public Vectorized16<float16x8_t, c10::Half, BlendH
...
+      value_type val7)
+      : Vectorized16(float16x8_t{
+            val0,
+            val1,
+            val2,
+            val3,
+            val4,
+            val5,
+            val6,
+            val7}) {}
```
Comment (malfet) [inline aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h:125]: This feels unrelated to PR title, isn't it?

**Review by malfet — APPROVED** [review]
LGTM if it builds

File: aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h (hunk 2 above)
Comment (swolchok) [reply aten/src/ATen/cpu/vec/vec128/vec128_half_neon.h:125]: folding back into previous rev. there was a problem with this constructor for bfloat16 so I meant to remove it from Vectorized16
